# Conceptual part

### A.1.

**Correct answer:**
ii. For a fixed value of IQ and GPA, college graduates earn more, on average, than high school graduates.

**Reason:** There are no further information about how the interactions (i.e. $X4$ and $X5$) does look like. So I have to assume something:

Looking at question 3 I assume $X4=\frac{X1}{X2}$. Since $X3 \in [0,1]$, I would not assume $X5=\frac{X1}{X3}$, because this could end in an devision by zero. Therefore I assume it the other way around, i.e. $X5=\frac{X3}{X1}$, (assuming that there is no data with a GPA < 2.0, because otherwise they would have been no graduates). Also the interaction $X5=X1*X3$ makes no sense, because this would result in a smaller salary when having a better GPA. (So I assume a higher salay for college graduates when having a higher GPA)




Since GPA (X1) and IQ (X2) are fixed, X4 is also fixed. So the statement depends on X3 and X5, which means that we have to show that there is a minimal value for $X1 \geq 2.0$ so that the part of the estimatied function containing the relevant predictors in our case, i.e. X3 and X5, is greater than 0, because both terms would be zero with my assumptions when having a high school graduate:
$$
\hat{\beta}_3*X3 + \hat{\beta}_5\frac{X3}{X1} > 0 \text{ | insert slopes and predictors}
\\
35*1-10\frac{1}{X1} > 0 \text{ | } +\frac{10}{X1}
\\
35 > \frac{10}{X1} \text{ | } * X1
\\
35X1 > 10 \text{ | } / 35
\\
X1 > \frac{10}{35}
$$
So the minimal value for the GPA would be $\frac{10}{35}=0.29$. But considering the domain here, it is impossible to graduate with a GPA < 2.0. So a college graduate is estimated to earn more than a high school graduate either way. (based on my assumptions above)




### A.2.
$$
\hat{y} = \hat{\beta}_0 + \hat{\beta}_1X1 + \hat{\beta}_2X2 + \hat{\beta}_3X3 + \hat{\beta}_4X4 + \hat{\beta}_5X5 \text{ | insert estimated interception and slope}
\\
\hat{y} = 50 + 20X1 + 0.07X2 + 35X3 + 0.01X4 + (-10)X5 \text{ | insert predictor values, with } X4=\frac{X1}{X2} \text{ and } X5=\frac{X3}{X1}
\\
\hat{y} = 50 + 20*4.0 + 0.07*110 + 35*1 + 0.01*\frac{4.0}{110}  -10 *\frac{1}{4.0} \text{ | calculate}
\\
\hat{y} = 170.204
$$
So the predicted salary for the given predictors is 170204\$ (based on my assumptions from A.1.).



### A.3.
**False:** The (regression) coefficient was estimated for the estimation of a function that estimates the salary based on specific data. It does not tell anything about the interaction/correlation between the predictors used in this estimation, even if it is very small. If we want to find evidence for the interaction between two predictors, we need to consider the correletion coefficient of the two predictors (i.e. GPA and IQ). Even a look at the t-statistic and the p-value would give only information about the dependency between the interaction of the predictors itself and the estimated outcome of the estimates function, but not about the interaction between theses predictors.


# Practical part (Assigment 3)

## I: Load data

In [1]:
import pandas

df = pandas.read_csv('/media/home/ngundermann/workspace/MachineLearning/data/ISLR/data/Boston.csv', index_col=0)

## I: Check again the accuracy of the linear regression

In [2]:
from ISLP.models import ModelSpec as MS
import statsmodels.api as sm

X = MS(['lstat', 'rm', 'nox', 'dis', 'ptratio']).fit_transform(df)
y = df['medv']
lm1 = sm.OLS(y,X).fit()
print(lm1.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.708
Model:                            OLS   Adj. R-squared:                  0.705
Method:                 Least Squares   F-statistic:                     242.6
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          3.67e-131
Time:                        09:12:23   Log-Likelihood:                -1528.7
No. Observations:                 506   AIC:                             3069.
Df Residuals:                     500   BIC:                             3095.
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept     37.4992      4.613      8.129      0.0

## I: Add interaction terms

In [3]:
df['lstat*rm'] = df['lstat']*df['rm']

X = MS(['lstat', 'rm', 'nox', 'dis', 'ptratio', 'lstat*rm']).fit_transform(df)
y = df['medv']
lm2 = sm.OLS(y,X).fit()
print(lm2.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.778
Model:                            OLS   Adj. R-squared:                  0.775
Method:                 Least Squares   F-statistic:                     290.8
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          2.48e-159
Time:                        09:12:23   Log-Likelihood:                -1459.9
No. Observations:                 506   AIC:                             2934.
Df Residuals:                     499   BIC:                             2963.
Df Model:                           6                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept      3.1518      4.880      0.646      0.5

## I: Interprete the results.

The interaction term of lstat and rm (lstat*rm) leads to an overall improvement of the fitted function, which can be shown by the higher R² value. According to the p-value the coefficient of the interaction term seems to be significant (p < 5%).

## I: Apply non-linear transformations to some predictors

In [4]:
df['(lstat*rm)2'] = df['lstat*rm']*df['lstat*rm']

X = MS(['lstat', 'rm', 'nox', 'dis', 'ptratio', '(lstat*rm)2', 'lstat*rm']).fit_transform(df)
y = df['medv']
lm3 = sm.OLS(y,X).fit()
print(lm3.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.781
Model:                            OLS   Adj. R-squared:                  0.778
Method:                 Least Squares   F-statistic:                     253.9
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          8.05e-160
Time:                        09:12:23   Log-Likelihood:                -1455.8
No. Observations:                 506   AIC:                             2928.
Df Residuals:                     498   BIC:                             2961.
Df Model:                           7                                         
Covariance Type:            nonrobust                                         
                  coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------
intercept      10.5522      5.499      1.919      

## I: Use ANOVA to check if the quadratic fit is superior to the linear fit.

In [5]:
from statsmodels.stats.anova import anova_lm

anova_lm(lm2, lm3)

df_resid          ssr  df_diff     ss_diff         F    Pr(>F)
0     499.0  9500.381881      0.0         NaN       NaN       NaN
1     498.0  9348.435955      1.0  151.945925  8.094303  0.004623

## I: Interprete the results

According to the linear fit, this model reached a R²-value of 0.781, which is the best so far. It also seems that all used predictors are useful, since the p-values of the t-statistics are all lower than 5%.

The output of the anova test shows the two models in realtion to the previous model. Since there is no previous model of the first model, the comparing columns (ssr, df_diff, ss_diff, F and Pr(>F)) are NaN or zero. In order to compare the two models. we have to look at these columns in the second line. Here the second model with the non linear predictor shows a better RSS since it is approx. 152 points lower than in the first model. Also the F statistic is a little bit higher and this increase is significant, since the p-value is lower than 5% (last column). So the non linear model seems to be the better fit here.

## I: Check if including additional polynomial terms, up to order, lead to an improvement in the model fit.

In [6]:
n = 5
for i in range(2,n+1):
    df[f'(lstat){i}'] = df['lstat'] ** i

X = MS(['lstat', 'rm', 'nox', 'dis', 'ptratio', '(lstat)2', '(lstat)3', '(lstat)4', '(lstat)5', 'lstat*rm']).fit_transform(df)
y = df['medv']
lm4 = sm.OLS(y,X).fit()
print(lm4.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.792
Model:                            OLS   Adj. R-squared:                  0.787
Method:                 Least Squares   F-statistic:                     188.0
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          1.80e-161
Time:                        09:12:23   Log-Likelihood:                -1443.5
No. Observations:                 506   AIC:                             2909.
Df Residuals:                     495   BIC:                             2956.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept     33.9246      7.663      4.427      0.0

## I: Use ANOVA to check if the quadratic fit is superior to the linear fit.

In [7]:
anova_lm(lm2, lm4)

df_resid          ssr  df_diff     ss_diff         F    Pr(>F)
0     499.0  9500.381881      0.0         NaN       NaN       NaN
1     495.0  8903.772453      4.0  596.609428  8.292038  0.000002

## I: Interprete the results

According to the linear fit, this model reached a R²-value of 0.792, which is the best so far. It also seems that all used predictors are useful, since the p-values of the t-statistics are all lower than 5%.

The output of the avona test shows the two models in realtion to the previous model. Since there is no previous model of the first model, the comparing columns (ssr, df_diff, ss_diff, F and Pr(>F)) are NaN or zero. In order to compare the two models, we have to look at these columns in the second line. Here the second model with the polynomial predictor shows a better RSS since it is approx. 597 points lower than in the linar model. Also the F statistic is a little bit higher and this increase is significant, since the p-value is lower than 5% (last column). So the polynomial model seems to be the better fit here.

## I: Fit and assess other non-linear transformations

In [8]:
import numpy as np

df['log(rm)'] = np.log(df['rm'])

X = MS(['lstat', 'rm', 'nox', 'dis', 'ptratio', '(lstat)2', '(lstat)3', '(lstat)4', '(lstat)5', 'log(rm)']).fit_transform(df)
y = df['medv']
lm5 = sm.OLS(y,X).fit()
print(lm5.summary())

                            OLS Regression Results                            
Dep. Variable:                   medv   R-squared:                       0.804
Model:                            OLS   Adj. R-squared:                  0.800
Method:                 Least Squares   F-statistic:                     202.6
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          7.10e-168
Time:                        09:12:23   Log-Likelihood:                -1428.4
No. Observations:                 506   AIC:                             2879.
Df Residuals:                     495   BIC:                             2925.
Df Model:                          10                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
intercept    172.9866     13.954     12.397      0.0

In [9]:
anova_lm(lm2, lm5)

df_resid          ssr  df_diff     ss_diff          F        Pr(>F)
0     499.0  9500.381881      0.0         NaN        NaN           NaN
1     495.0  8386.756361      4.0  1113.62552  16.431997  1.190966e-12

## I: Interprete the results

According to the linear fit, this model reached a R²-value of 0.804, which is the best so far. It also seems that all used predictors are useful, since the p-values of the t-statistics are all lower than 5%.


The output of the anova test shows the two models in realtion to the previous model. Since there is no previous model of the first model, the comparing columns (ssr, df_diff, ss_diff, F and Pr(>F)) are NaN or zero. In order to compare the two models, we have to look at these columns in the second line. Here the second model with the log predictor shows a better RSS since it is approx. 1114 points lower than in the linar model. Also the F statistic is a little bit higher and this increase is significant, since the p-value is lower than 5% (last column). So the model with the log predictor seems to be the better fit here.

## I: Use categorical predictors

## I: Load data

In [10]:
df = pandas.read_csv('/media/home/ngundermann/workspace/MachineLearning/data/ISLR/data/Carseats.csv', index_col=0)
for column in df.columns.values:
    print(f'{df[column].describe()}\n')

count    400.000000
mean       7.496325
std        2.824115
min        0.000000
25%        5.390000
50%        7.490000
75%        9.320000
max       16.270000
Name: Sales, dtype: float64

count    400.000000
mean     124.975000
std       15.334512
min       77.000000
25%      115.000000
50%      125.000000
75%      135.000000
max      175.000000
Name: CompPrice, dtype: float64

count    400.000000
mean      68.657500
std       27.986037
min       21.000000
25%       42.750000
50%       69.000000
75%       91.000000
max      120.000000
Name: Income, dtype: float64

count    400.000000
mean       6.635000
std        6.650364
min        0.000000
25%        0.000000
50%        5.000000
75%       12.000000
max       29.000000
Name: Advertising, dtype: float64

count    400.000000
mean     264.840000
std      147.376436
min       10.000000
25%      139.000000
50%      272.000000
75%      398.500000
max      509.000000
Name: Population, dtype: float64

count    400.000000
mean     115.795000

## I: For the categorical predictors, generate dummy variables

In [11]:
df['Urban_dummy'] = np.where(df['Urban'] == 'Yes', 1, 0)
df['US_dummy'] = np.where(df['US'] == 'Yes', 1, 0)
df['ShelveLoc_Good_dummy'] = np.where(df['ShelveLoc'] == 'Good', 1, 0)
df['ShelveLoc_Medium_dummy'] = np.where(df['ShelveLoc'] == 'Medium', 1, 0)

In [12]:
rel_columns = list(filter((lambda name: name != 'Sales' and name != 'US' and name != 'Urban' and name != 'ShelveLoc'), df.columns.values))
X = MS(rel_columns).fit_transform(df)
y = df['Sales']
lm6 = sm.OLS(y,X).fit()
print(lm6.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.873
Model:                            OLS   Adj. R-squared:                  0.870
Method:                 Least Squares   F-statistic:                     243.4
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          1.60e-166
Time:                        09:12:24   Log-Likelihood:                -568.99
No. Observations:                 400   AIC:                             1162.
Df Residuals:                     388   BIC:                             1210.
Df Model:                          11                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
intercept                  5

## I: Interprete the results

The fitted model is an overall good model, since the R² value is quite high at 0.873. Anyway there are some predictors that might be irrelevant, since the p-value if higher than 5%, i.e. US_dummy, Urban_dummy, Education and population. So it seems that the US predictor, the Urban predictor, the Population and the education does not play a role according to the prediction. 

## I: Play around with the predictors

In [13]:
irrelevant_columns = ['Sales', 'US', 'Urban', 'ShelveLoc', 
                      'Education', 'Population', 'US_dummy', 'Urban_dummy' ]

df['Income*CompPrice'] = df['Income'] * df['CompPrice']
df['Income*Advertising'] = df['Income'] * df['Advertising']

rel_columns = list(filter((lambda name: name not in irrelevant_columns), df.columns.values))

X = MS(rel_columns).fit_transform(df)
y = df['Sales']
lm7 = sm.OLS(y,X).fit()
print(lm7.summary())

                            OLS Regression Results                            
Dep. Variable:                  Sales   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.875
Method:                 Least Squares   F-statistic:                     310.0
Date:                Fri, 02 Feb 2024   Prob (F-statistic):          1.15e-171
Time:                        09:12:24   Log-Likelihood:                -562.67
No. Observations:                 400   AIC:                             1145.
Df Residuals:                     390   BIC:                             1185.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                             coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------------
intercept                  2

Beaten! (by 0.007 R²)